modules/core/src/main/java/org/apache/ignite/internal/processors/continuous/GridContinuousProcessor.java

```diff
@@ -660,13 +660,34 @@ private void registerHandlerOnJoin(UUID srcNodeId, UUID routineId, IgnitePredica
                 ctx.resource().injectGeneric(nodeFilter);
 
             if (nodeFilter == null || nodeFilter.apply(ctx.discovery().localNode())) {
-                    registerHandler(srcNodeId,
+                registerHandler(srcNodeId,
                         routineId,
                         hnd,
                         bufSize,
                         interval,
                         autoUnsubscribe,
                         false);
+
+                if (ctx.config().isPeerClassLoadingEnabled()) {
+                    // Peer class loading cannot be performed before a node joins, so we delay the deployment.
+                    // Run the deployment task in the system pool to avoid blocking of the discovery thread.
+                    ctx.discovery().localJoinFuture().listen(f -> ctx.closure().runLocalSafe(new GridPlainRunnable() {
+                        @Override public void run() {
+                            try {
+                                hnd.p2pUnmarshal(srcNodeId, ctx);
+                            }
+                            catch (IgniteCheckedException | IgniteException e) {
+                                U.error(log, "Failed to unmarshal continuous routine handler [" +
+                                    "routineId=" + routineId +
+                                    ", srcNodeId=" + srcNodeId + ']', e);
+
+                                ctx.failure().process(new FailureContext(FailureType.CRITICAL_ERROR, e));
+
+                                unregisterHandler(routineId, hnd, false);
```

Comment:
Why is FailureHandler called before unregistering the CQ handler?
I don't think, this is an issue, but could you please clarify?